#!/bin/bash
#Quantification pipeline
#Salmon results

RSEM(){

  filename=$1

  #Start the clock for RSEM
  start_RSEM=`date +%s`

  #RSEM
  ./Simulation/RSEM-1.3.0/rsem-calculate-expression --paired-end --star\
		    --star-path Simulation/STAR/bin/Linux_x86_64 \
		    -p 8 \
        --append-names \
		    --single-cell-prior --calc-pme \
		    --time \
        Simulation/data/simulated/$filename'_1.fq' Simulation/data/simulated/$filename'_2.fq' \
        Simulation/ref/reference Simulation/RSEM_results/$filename


	#Stop the clock for RSEM
	stop_RSEM=`date +%s`

	#Delete everything generated by RSEM except isoforms.results
	rm -r Simulation/RSEM_results/$filename'.stat'
	rm -r Simulation/RSEM_results/$filename'.transcript'*
	rm -r Simulation/RSEM_results/$filename'.genome'*
	rm -r Simulation/RSEM_results/$filename'.genes.results'

  #Trim the text added to the transcript names in the results file
  python ./trim.py `pwd` Simulation/RSEM_results/ $filename'.isoforms.results'
  mv Simulation/RSEM_results/'trimmed'$filename'.isoforms.results' Simulation/RSEM_results/$filename'.isoforms.results'

  #Sort the results file
  sort -n -k1.8 Simulation/RSEM_results/$filename'.isoforms.results' > Simulation/RSEM_results/$filename'.sortedisoforms.results'
  mv Simulation/RSEM_results/$filename'.sortedisoforms.results' Simulation/RSEM_results/$filename'.isoforms.results'

	printf $filename","$((stop_RSEM-start_RSEM))"\n">> Simulation/time_stats/time_RSEM.csv
}

Cufflinks(){

  #Rename/reformat input arguments
  filename=$1
  gtf=$2
  library=$3

  mkdir Simulation/Cufflinks_results/$filename

  if [ ! -f Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.out.bam' ]; then
    STAR $filename
  fi

  #Start the clock for cufflinks
  start_Cufflinks=`date +%s`

  ./Simulation/cufflinks-2.2.1.Linux_x86_64/cufflinks -p 8 -F 0 --library-type $library -G $gtf -o Simulation/Cufflinks_results/$filename Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.out.bam'

  stop_cufflinks=`date +%s`

  printf $filename","$((stop_cufflinks-start_cufflinks))"\n">> Simulation/time_stats/time_cufflinks.csv

}



Salmon(){

	#Rename/reformat input arguments
  filename=$1

	mkdir Simulation/Salmon_results/Salmon_Alignment_Results/$filename
	mkdir Simulation/Salmon_results/Salmon_SMEM_results/$filename
	mkdir Simulation/Salmon_results/Salmon_quasi_results/$filename

  if [ ! -f Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.out.bam' ]; then
    STAR $filename
  fi

	#Start the clock for Salmon alignment mode
	start_Salmon_align=`date +%s`

	#Salmon alignment mode
	Simulation/Salmon-0.7.2_linux_x86_64/bin/salmon --no-version-check quant -a Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.out.bam' -t Simulation/ref/reference.transcripts.fa -l A -o Simulation/Salmon_results/Salmon_Alignment_Results/$filename -p 8

	#Stop the clock for Salmon alignment mode

	stop_Salmon_align=`date +%s`

	rm -r Simulation/Salmon_results/Salmon_Alignment_Results/$filename/aux_info
	rm -r Simulation/Salmon_results/Salmon_Alignment_Results/$filename/cmd_info.json
	rm -r Simulation/Salmon_results/Salmon_Alignment_Results/$filename/lib_format_counts.json
	rm -r Simulation/Salmon_results/Salmon_Alignment_Results/$filename/libParams
	rm -r Simulation/Salmon_results/Salmon_Alignment_Results/$filename/logs

  #Sort the results file
  sort -n -k1.8 Simulation/Salmon_results/Salmon_Alignment_Results/$filename/quant.sf > Simulation/Salmon_results/Salmon_Alignment_Results/$filename/quantsorted.sf
  mv Simulation/Salmon_results/Salmon_Alignment_Results/$filename/quantsorted.sf Simulation/Salmon_results/Salmon_Alignment_Results/$filename/quant.sf

  #If the SMEM index doesn't exist, make it
  if [ ! "$(ls -A Simulation/indices/Salmon_SMEM)" ]; then
    start_Salmon_SMEM_index=`date +%s`
    Simulation/Salmon-0.7.2_linux_x86_64/bin/salmon index -t Simulation/ref/reference.transcripts.fa -i Simulation/indices/Salmon_SMEM/transcripts_index_SMEM --type fmd -p 8
    stop_Salmon_SMEM_index=`date +%s`
    printf $filename","$((stop_Salmon_SMEM_index-start_Salmon_SMEM_index)) >> Simulation/time_stats/time_Salmon_SMEM_index.csv
  fi

	#Start the clock for Salmon alignment free SMEM
	start_Salmon_SMEM=`date +%s`

	#Salmon alignment free SMEM
	Simulation/Salmon-0.7.2_linux_x86_64/bin/salmon --no-version-check quant -i Simulation/indices/Salmon_SMEM/transcripts_index_SMEM -l A -1 Simulation/data/simulated/$filename'_1.fq' -2 Simulation/data/simulated/$filename'_2.fq' -o Simulation/Salmon_results/Salmon_SMEM_results/$filename -p 8

	#Stop the clock for Salmon SMEM
	stop_Salmon_SMEM=`date +%s`

	rm -r Simulation/Salmon_results/Salmon_SMEM_results/$filename/aux_info
	rm -r Simulation/Salmon_results/Salmon_SMEM_results/$filename/cmd_info.json
	rm -r Simulation/Salmon_results/Salmon_SMEM_results/$filename/lib_format_counts.json
	rm -r Simulation/Salmon_results/Salmon_SMEM_results/$filename/libParams
	rm -r Simulation/Salmon_results/Salmon_SMEM_results/$filename/logs

  #Sort the results file
  sort -n -k1.8 Simulation/Salmon_results/Salmon_SMEM_results/$filename/quant.sf > Simulation/Salmon_results/Salmon_SMEM_results/$filename/quantsorted.sf
  mv Simulation/Salmon_results/Salmon_SMEM_results/$filename/quantsorted.sf Simulation/Salmon_results/Salmon_SMEM_results/$filename/quant.sf

  #If the quasi index doesn't exist, make it
  if [ ! "$(ls -A Simulation/indices/Salmon_quasi)" ]; then
    start_Salmon_quasi_index=`date +%s`
    Simulation/Salmon-0.7.2_linux_x86_64/bin/salmon index -t Simulation/ref/reference.transcripts.fa -i Simulation/indices/Salmon_quasi/transcripts_index_quasi --type quasi -k 31 -p 8
    stop_Salmon_quasi_index=`date +%s`
    printf $filename","$((stop_Salmon_quasi_index-start_Salmon_quasi_index)) >> Simulation/time_stats/time_Salmon_quasi_index.csv
  fi

	#Start the clock for alignment free quasi
	start_Salmon_quasi=`date +%s`

	#Salmon alignment free quasi
	Simulation/Salmon-0.7.2_linux_x86_64/bin/salmon --no-version-check quant -i Simulation/indices/Salmon_quasi/transcripts_index_quasi -l A -1 Simulation/data/simulated/$filename'_1.fq' -2 Simulation/data/simulated/$filename'_2.fq' -o Simulation/Salmon_results/Salmon_quasi_results/$filename -p 8

	#Stop the clock for other alignment quasi
	stop_Salmon_quasi=`date +%s`

	rm -r Simulation/Salmon_results/Salmon_quasi_results/$filename/aux_info
	rm -r Simulation/Salmon_results/Salmon_quasi_results/$filename/cmd_info.json
	rm -r Simulation/Salmon_results/Salmon_quasi_results/$filename/lib_format_counts.json
	rm -r Simulation/Salmon_results/Salmon_quasi_results/$filename/libParams
	rm -r Simulation/Salmon_results/Salmon_quasi_results/$filename/logs

  #Sort the results file
  sort -n -k1.8 Simulation/Salmon_results/Salmon_quasi_results/$filename/quant.sf > Simulation/Salmon_results/Salmon_quasi_results/$filename/quantsorted.sf
  mv Simulation/Salmon_results/Salmon_quasi_results/$filename/quantsorted.sf Simulation/Salmon_results/Salmon_quasi_results/$filename/quant.sf

	#Output all times into Time.csv
	printf $filename","$((stop_Salmon_align-start_Salmon_align))","$((stop_Salmon_SMEM-start_Salmon_SMEM))","$((stop_Salmon_quasi-start_Salmon_quasi))"\n" >> Simulation/time_stats/time_Salmon.csv

}

eXpress () {

  filename=$1

	#make a directory for the results of eXpress for each cell
	mkdir Simulation/eXpress_results/$filename

  if [ ! -f Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.out.bam' ]; then
    STAR $filename
  fi

	#Start the clock for eXpress
	start_eXpress=`date +%s`

	#Run eXpress
	./Simulation/express-1.5.1-linux_x86_64/express Simulation/ref/reference.transcripts.fa Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.out.bam' -o Simulation/eXpress_results/$filename

	#Stop the clock for eXpress
	stop_eXpress=`date +%s`

  #Remove first column of results file then sort the file
  cut -d$'\t' -f 2- Simulation/eXpress_results/$filename/results.xprs > Simulation/eXpress_results/$filename/resultscut.xprs
  sort -n -k1.8 Simulation/eXpress_results/$filename/resultscut.xprs > Simulation/eXpress_results/$filename/results.xprs

	printf $filename","$((stop_eXpress-start_eXpress))"\n" >> $TEAM/time_eXpress.csv

}

Kallisto () {

	#make a directory for the results of eXpress for each cell
	filename=$1
	mkdir Simulation/Kallisto_results/$filename

  #If there is no index for kallisto, make it
  if [ ! "$(ls -A Simulation/indices/Kallisto)" ]; then
    start_kallisto_index=`date +%s`
    ./Simulation/kallisto_linux-v0.43.0/kallisto index -i Simulation/indices/Kallisto/transcripts.idx Simulation/ref/reference.transcripts.fa
    stop_kallisto_index=`date +%s`
    printf $filename","$((stop_kallisto_index-start_kallisto_index)) >> Simulation/time_stats/time_kallisto_index.csv
  fi

	#Start the clock for kallisto
	start_kallisto=`date +%s`

	./Simulation/kallisto_linux-v0.43.0/kallisto quant -i Simulation/indices/Kallisto/transcripts.idx --threads=8 --output-dir=Simulation/Kallisto_results/$filename Simulation/data/simulated/$filename'_1.fq' Simulation/data/simulated/$filename'_2.fq'

	#Stop the clock for kallisto
	stop_kallisto=`date +%s`

	printf $filename","$((stop_kallisto-start_kallisto))"\n" >> $TEAM/time_kallisto.csv

  #Sort the results file
  sort -n -k1.8 Simulation/Kallisto_results/$filename/abundance.tsv > Simulation/Kallisto_results/$filename/abundancesorted.tsv
  mv Simulation/Kallisto_results/$filename/abundancesorted.tsv Simulation/Kallisto_results/$filename/abundance.tsv

}

Sailfish(){

  #make a directory for the results of sailfish for each cell
  filename=$1
  library_type=$2
  mkdir Simulation/Sailfish_results/$filename
  export LD_LIBRARY_PATH=`pwd`/Simulation/Sailfish-0.6.3-Linux_x86-64/lib:$LD_LIBRARY_PATH
  export PATH=`pwd`/Simulation/Sailfish-0.6.3-Linux_x86-64/bin:$PATH

  #If there is no index for sailfisg, make it
  if [ ! "$(ls -A Simulation/indices/Kallisto)" ]; then
    start_sailfish_index=`date +%s`
    ./Simulation/Sailfish-0.6.3-Linux_x86-64/bin/sailfish index -p 8 -t Simulation/ref/reference.transcripts.fa -o Simulation/indices/Sailfish/ -k 31
    stop_sailfish_index=`date +%s`
    printf $filename","$((stop_sailfish_index-start_sailfish_index)) >> Simulation/time_stats/time_sailfish_index.csv
  fi

  #Start the clock for sailfish
  start_sailfish=`date +%s`

  echo ./Simulation/Sailfish-0.6.3-Linux_x86-64/bin/sailfish quant -p 8 -i Simulation/indices/Sailfish/ -l "$library_type" { --mates1 Simulation/data/simulated/$filename"_1.fq" --mates2 Simulation/data/simulated/$filename"_2.fq"} -o Simulation/Sailfish_results/$filename

  ./Simulation/Sailfish-0.6.3-Linux_x86-64/bin/sailfish quant -p 8 -i Simulation/indices/Sailfish/ -l "T=PE:O=><:S=U" { --mates1 Simulation/data/simulated/$filename"_1.fq" --mates2 Simulation/data/simulated/$filename"_2.fq"} -o Simulation/Sailfish_results/$filename

  stop_sailfish=`date +%s`

  printf $filename","$((stop_sailfish-start_sailfish))"\n" >> $TEAM/time_sailfish.csv

  rm Simulation/Sailfish_results/$filename/logs
  rm Simulation/Sailfish_results/$filename/quant_bias_corrected.sf
  rm Simulation/Sailfish_results/$filename/reads.count_info
  rm Simulation/Sailfish_results/$filename/reads.sfc

  echo "Transcript      Length  TPM     RPKM    KPKM    EstimatedNumKmers       EstimatedNumReads" >> Simulation/Sailfish_results/$filename/quantsorted.sf
  sed 's/^# *//' Simulation/Sailfish_results/$filename/quant.sf | tail -n +6 | sort -n -k1.8 >> Simulation/Sailfish_results/$filename/quantsorted.sf
  mv Simulation/Sailfish_results/$filename/quantsorted.sf Simulation/Sailfish_results/$filename/quant.sf


}

STAR(){

  filename=$1

	#Start the clock for STAR
	start_STAR=`date +%s`

	#Make STAR reference
	./Simulation/STAR/bin/Linux_x86_64/STAR --runThreadN 8 --genomeDir Simulation/ref --readFilesIn Simulation/data/simulated/$filename'_1.fq' Simulation/data/simulated/$filename'_2.fq' --outFileNamePrefix Simulation/bamfiles/simulated/$filename'XS' --outSAMstrandField intronMotif --outSAMtype BAM Unsorted --quantMode TranscriptomeSAM

	#Stop the clock for STAR
	stop_STAR=`date +%s`

	#Start the clock for Samtools sort
	start_samtools_sort=`date +%s`

	#Sort STAR bamfiles
	./Simulation/samtools-1.3.1/samtools sort Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.out.bam' -T Simulation/bamfiles/simulated/$filename'temp' -o Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.sortedByCoord.out.bam'


	#Stop the clock for Samtools sort
	stop_samtools_sort=`date +%s`

	#Start the clock for Samtools index
	start_samtools_index=`date +%s`

	#Index STAR bamfiles
	./Simulation/samtools-1.3.1/samtools index Simulation/bamfiles/simulated/$filename'XSAligned.toTranscriptome.sortedByCoord.out.bam'

	#Stop the clock for Samtools
	stop_samtools_index=`date +%s`

	printf $filename","$((stop_STAR-start_STAR))","$((stop_samtools_sort-start_samtools_sort))","$((stop_samtools_index-start_samtools_index))"\n" >> Simulation/time_stats/time_STARXS_samtools.csv

}


#Export functions
export -f RSEM
export -f Salmon
export -f eXpress
export -f Kallisto
export -f STAR
export -f Cufflinks
export -f Sailfish

"$@"
